### PR TITLE
Add Tabnet support

### PR DESCRIPTION
**Pull request by TZTsai**
## Description
I have added the TabNet model now. It is added as a building block in other models, similar to MLP. To incorporate the new model in current plugins like tabular_gan, tabddpm, etc., My approach is to remove the hyper-parameters specific to the MLP model and replace them with two parameters `model_type` and `model_params`. I have created a file "factory.py" that contains functions that given a name string and a dict of params, will return an nn module, or an activation layer, or a feature encoder. The `get_model` function, provided with `model_type` and `model_params`, can instantiate an NN block such as an `MLP`, `TransformerModel`, or `TabNet`. Consequently, by substituting MLP blocks with the dynamically instantiated blocks, we can provide more flexibility to the present plugins. I have tried using `TabNet` as the diffusion model of TabDDPM, instead of `MLP`, and it can successfully fit and generate data, although the performance is acutally not as good as MLP.

## Checklist
- [X] I have followed the [Contribution Guidelines](https://github.com/vanderschaarlab/.github/blob/main/CONTRIBUTING.md) and [Code of Conduct](https://github.com/vanderschaarlab/.github/blob/main/CODE_OF_CONDUCT.md)
- [x] I have commented my code following the [van der Schaar Lab Styleguide](https://github.com/vanderschaarlab/.github/blob/main/STYLEGUIDE.md)
- [X] I have labelled this PR with the relevant [Type labels](https://github.com/vanderschaarlab/.github/labels?q=Type%3A)
- [X] My changes are covered by tests

### CODE REVIEW

File: src/synthcity/plugins/core/models/tabnet.py [hunk 1]
```diff
@@ -0,0 +1,831 @@
...
+        )
+
+    def forward(self, x: torch.Tensor) -> torch.Tensor:
+        x = self.fc(x)
+        x = self.bn(x)
+        out = torch.mul(x[:, : self.output_dim], torch.sigmoid(x[:, self.output_dim :]))
+        return out
+
+
+# class TabNetDecoder(torch.nn.Module):
```
Comment (bcebere) [inline src/synthcity/plugins/core/models/tabnet.py:514]: Please remove these comments if the code is not used

**Review by bcebere — COMMENTED** [review]
Great work @TZTsai ! Thank you for another great PR! 

I left some questions

File: src/synthcity/plugins/core/models/tabnet.py (hunk 1 above)
Comment (TZTsai) [reply src/synthcity/plugins/core/models/tabnet.py:514]: Done.

File: src/synthcity/utils/dataframe.py [hunk 2]
```diff
@@ -1,7 +1,19 @@
+# stdlib
+import enum
+
 # third party
 import pandas as pd
 
 
+class TaskType(enum.Enum):
```
[the remark this answers is not in the document]
Comment (TZTsai) [reply src/synthcity/utils/dataframe.py:8]: Actually they are not useful now. I have removed it.

**Comment by TZTsai** [comment]
I don't understand why there are errors of `TypeError: NDFrame.describe() got an unexpected keyword argument 'datetime_is_numeric'` and `AttributeError: 'DPOptimizer' object has no attribute '_optimizer_step_pre_hooks'. Did you mean: '_optimizer_step_code'?` in some tests in "tests/plugins/survival_analysis/test_survae.py". I ran these tests locally (Windows) and they were all passed. I searched in the project but did not find any usage of `DataFrame.describe()` or `DPOptimizer`. It seems not due to the updates I made, or is it?

**Comment by DrShushen** [comment]
> I don't understand why there are errors of `TypeError: NDFrame.describe() got an unexpected keyword argument 'datetime_is_numeric'` and `AttributeError: 'DPOptimizer' object has no attribute '_optimizer_step_pre_hooks'. Did you mean: '_optimizer_step_code'?` in some tests in "tests/plugins/survival_analysis/test_survae.py". I ran these tests locally (Windows) and they were all passed. I searched in the project but did not find any usage of `DataFrame.describe()` or `DPOptimizer`. It seems due to the lack of test support for Pandas 2.0 and PyTorch 2.0, is it correct?

I'm guessing `TypeError: NDFrame.describe() got an unexpected keyword argument 'datetime_is_numeric` is a problem with `pandas 2.0.0` compatibility, similar to this: https://github.com/CamDavidsonPilon/lifelines/issues/1512

**Comment by bcebere** [comment]
Please merge the `main` branch into this, the test should pass after that.

File: setup.cfg [hunk 3]
```diff
@@ -34,8 +34,8 @@ python_requires = >=3.7
 install_requires =
     scikit-learn>=1.0
     nflows>=0.14
-    pandas>=1.3,<2.0
-    torch>=1.10,<2.0
+    pandas>=2.0
```
Comment (bcebere) [inline setup.cfg:37]: Please don't change the versioning for torch and pandas for now, there are other components affected.

**Review by bcebere — CHANGES_REQUESTED** [review]

File: setup.cfg (hunk 3 above)
Comment (TZTsai) [reply setup.cfg]: I have updated

**Review by bcebere — APPROVED** [review]
LGTM! 
@robsdavis please review as well.

**Comment by robsdavis** [comment]
Thanks for you work on this @TZTsai! I'll do my review and look at merging, if you can get the tests to pass again. Thanks so much!

**Comment by TZTsai** [comment]
> Thanks for you work on this @TZTsai! I'll do my review and look at merging, if you can get the tests to pass again. Thanks so much!

Hi @robsdavis! I have updated the code but somehow a test failed to import torch. I reviewed the changes but none of them seems likely to be the cause of this problem. Could you please help me sort out the problem?

**Comment by TZTsai** [comment]
@bcebere @robsdavis There is an error when installing dependencies. Could you help check it?

**Comment by TZTsai** [comment]
> @bcebere @robsdavis There is an error when installing dependencies. Could you help check it?

I think a possible cause is the new Pip 23.1 released on April 15th.

**Review by robsdavis — APPROVED** [review]

**Pull request merged**